python/pyspark/streaming/tests.py

```diff
@@ -398,6 +399,69 @@ def func(dstream):
         self._test_func(input, func, expected)
 
 
+class StreamingListenerTests(PySparkStreamingTestCase):
+
+    duration = .5
+
+    class BatchInfoCollector(StreamingListener):
+
+        def __init__(self):
+            super(StreamingListener, self).__init__()
+            self.batchInfosCompleted = []
+            self.batchInfosStarted = []
+            self.batchInfosSubmitted = []
+
+        def onBatchSubmitted(self, batchSubmitted):
+            self.batchInfosSubmitted.append(self.getEventInfo(batchSubmitted))
+
+        def onBatchStarted(self, batchStarted):
+            self.batchInfosStarted.append(self.getEventInfo(batchStarted))
+
+        def onBatchCompleted(self, batchCompleted):
+            self.batchInfosCompleted.append(self.getEventInfo(batchCompleted))
+
+    def test_batch_info_reports(self):
+        batch_collector = self.BatchInfoCollector()
+        self.ssc.addStreamingListener(batch_collector)
+        input = [[1], [2], [3], [4]]
+
+        def func(dstream):
+            return dstream.map(int)
+        expected = [[1], [2], [3], [4]]
+        self._test_func(input, func, expected)
+
+        # Test occasionally fails without a delay
+        time.sleep(.1)
```

Comment:
The listener sometimes receives 3 batchCompleted events instead of 4 when I run the test, even though it always gets the correct job output. However, when I add a minor delay the test passes consistently. Can I get someone's opinion on this?